src/mesh.cpp

```diff
@@ -121,18 +148,62 @@ void MeshName::createDirectories() const
   }
 }
 
-void MeshName::save(const Mesh &mesh) const
+void MeshName::save(const Mesh &mesh, const std::string &dataname) const
 {
-  assert(mesh.positions.size() == mesh.data.size());
-  createDirectories();
-  std::ofstream out(filename(), std::ios::trunc);
-  out.precision(std::numeric_limits<long double>::max_digits10);
-  for (size_t i = 0; i < mesh.positions.size(); i++) {
-    out << mesh.positions[i][0] << " "
-        << mesh.positions[i][1] << " "
-        << mesh.positions[i][2] << " "
-        << mesh.data[i] << '\n';
+  const int                            numComp          = mesh.positions.size() / mesh.data.size();
+  vtkSmartPointer<vtkUnstructuredGrid> unstructuredGrid = vtkSmartPointer<vtkUnstructuredGrid>::New();
+  vtkSmartPointer<vtkPoints>           points           = vtkSmartPointer<vtkPoints>::New();
+  vtkSmartPointer<vtkDoubleArray>      data             = vtkDoubleArray::New();
+
+  data->SetName(dataname.c_str());
+  data->SetNumberOfComponents(numComp);
+
+  // Insert Points and Point Data
+  for (size_t i = 0; i < mesh.positions.size() - 1; i++) {
+    points->InsertNextPoint(mesh.positions[i][0], mesh.positions[i][1], mesh.positions[i][2]);
+    for (int j = 0; j < numComp; j++)
+      data->InsertNextTuple(&mesh.data[i * numComp + j]);
   }
+
+  unstructuredGrid->SetPoints(points);
+  unstructuredGrid->GetPointData()->AddArray(data);
+
+  // Connectivity Information
+
+  if (mesh.triangles.size() > 0) {
+    vtkSmartPointer<vtkCellArray> triArray = vtkSmartPointer<vtkCellArray>::New();
+    for (size_t i = 0; i <= mesh.triangles.size(); i++) {
```

Comment:
Why does this loop run until `<=`? With this, the number of inserted cells becomes `sieze+1`?

Reply:
Same for thr edges below

Reply:
That is a bug coming from my previous trial, there was a bug some triangles/edges were missing and somehow solved with this but right it should be `<` not `<=` for both.